Update Dynatrace README (#8)
dynatrace-oss-contrib/telegraf: plugins/outputs/dynatrace/README.md

# Dynatrace Output Plugin

> This plugin sends Telegraf metrics to [Dynatrace](https://www.dynatrace.com) via the [Dynatrace Metrics API V2](https://www.dynatrace.com/support/help/dynatrace-api/environment-api/metric-v2/). It may be run alongside the Dynatrace OneAgent for automatic authentication or it may be run standalone on a host without a OneAgent by specifying a URL and API Token.

Reviewer: ```suggestion
This plugin sends Telegraf metrics to [Dynatrace](https://www.dynatrace.com) via the [Dynatrace Metrics API V2](https://www.dynatrace.com/support/help/dynatrace-api/environment-api/metric-v2/). It may be run alongside the Dynatrace OneAgent for automatic authentication or it may be run standalone on a host without a OneAgent by specifying a URL and API Token.
More information on the plugin can be found in the [Dynatrace documentation](https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/ingestion-methods/telegraf/).
```

## Requirements

> You will either need a Dynatrace OneAgent (version 1.201 or higher) installed on the same host as Telegraf; or a Dynatrace environment with version 1.202 or higher. Monotonic counters (e.g. diskio.reads, system.uptime) require Dynatrace 208 or later.

Reviewer: ```suggestion
You will either need a Dynatrace OneAgent (version 1.201 or higher) installed on the same host as Telegraf; or a Dynatrace environment with version 1.202 or higher. Monotonic counters (e.g. `diskio.reads`, `system.uptime`) require Dynatrace 208 or later.
```

- Telegraf minimum version: Telegraf 1.16

> - Plugin minimum tested version: 1.16

Reviewer: ```suggestion
```
Doesn't this say the same thing?

## Getting Started

Setting up Telegraf is explained in the [Telegraf Documentation](https://docs.influxdata.com/telegraf/latest/introduction/getting-started/).

> The Dynatrace exporter may be enabled by adding an `[[outputs.dynatrace]]` section to your `telegraf.conf`.

Reviewer: ```suggestion
The Dynatrace exporter may be enabled by adding an `[[outputs.dynatrace]]` section to your `telegraf.conf` config file.
```

> All configurations are optional, but if a `url` is specified then an `api_token` is required.

Reviewer: ```suggestion
All configurations are optional, but if a `url` is specified, then an `api_token` is required.
```
?

Reviewer 2: Is it not possible to export to a OneAgent on a different port? Then the URI would be specified, but no token required.

Author: @arminru from https://owl.purdue.edu/owl/general_writing/punctuation/commas/extended_rules_for_commas.html

>However, don't put a comma after the main clause when a dependent (subordinate) clause follows it (except for cases of extreme contrast).

The original version is correct use of commas (english is the worst).

Reviewer: Good point, this is covered below somewhere but not consistent with this.

Author: @pirgeo I haven't seen any documentation that suggests it is possible. have you seen something like that?

Reviewer: https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/ingestion-methods/local-api/#communication-port
https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/ingestion-methods/telegraf/#communication-port

Author: Updated

To see all available options, see [Configuration](#configuration) below.

### Running alongside Dynatrace OneAgent

> If you run the Telegraf agent on a host or VM that is monitored by the Dynatrace OneAgent then you only need to enable the plugin, but need no further configuration. The Dynatrace telegraf output plugin will send all metrics to the OneAgent which will use its secure and load balanced connection to send the metrics to your Dynatrace SaaS or Managed environment.

Reviewer: ```suggestion
If you run the Telegraf agent on a host or VM that is monitored by the Dynatrace OneAgent then you only need to enable the plugin, but need no further configuration. The Dynatrace Telegraf output plugin will send all metrics to the OneAgent which will use its secure and load balanced connection to send the metrics to your Dynatrace SaaS or Managed environment.
Depending on your environment, you might have to enable Telegraf ingestion on the OneAgent first as described in the [Dynatrace documentation](https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/ingestion-methods/telegraf/).
```

Reviewer 2: `s/enable Telegraf ingestion/enable metrics ingestion` on line 22?

Author: ```suggestion
If you run the Telegraf agent on a host or VM that is monitored by the Dynatrace OneAgent then you only need to enable the plugin, but need no further configuration. The Dynatrace Telegraf output plugin will send all metrics to the OneAgent which will use its secure and load balanced connection to send the metrics to your Dynatrace SaaS or Managed environment.
Depending on your environment, you might have to enable metrics ingestion on the OneAgent first as described in the [Dynatrace documentation](https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/ingestion-methods/telegraf/).
```

> note: The name and identifier of the Telegraf host will be added as a dimension to every metric. If this is undesireable, then the output plugin may be used in standalone mode using the directions below.

Reviewer: ```suggestion
Note: The name and identifier of the host running Telegraf will be added as a dimension to every metric. If this is undesirable, then the output plugin may be used in standalone mode using the directions below.
```

```toml
[[outputs.dynatrace]]

>   ## No options are required. By default, metrics will be exported via the OneAgent on the local host.

Reviewer: Is the default export interval fine for us?

Author: The flush interval is not set per-output so we don't have a lot of control there. We could mention a recommended interval if we want. Having a hard time finding a documented default flush interval but it appears to be 10s according to some random internet posts.

Reviewer: Mentioning a recommended one sounds good. What did we say in the other exporters? I think it was 10-60s.

```

> ## Running standalone

Reviewer 2: ```suggestion
## Running in standalone mode
```

Author: I think what we have is plenty descriptive without making the heading unnecessarily long

If you run the Telegraf agent on a host or VM without a OneAgent you will need to configure the environment API endpoint to send the metrics to and an API token for security.

> You will also need to configure an API token for secure access. Find out how to create a token [here](https://www.dynatrace.com/support/help/dynatrace-api/environment-api/tokens/) or simply navigate to **Settings > Integration > Dynatrace API** in your Dynatrace environment and create a token with Dynatrace API and create a new token with 

Reviewer: ```suggestion
You will also need to configure an API token for secure access. Find out how to create a token in the [Dynatrace documentation](https://www.dynatrace.com/support/help/dynatrace-api/basics/dynatrace-api-authentication/) or simply navigate to **Settings > Integration > Dynatrace API** in your Dynatrace environment and create a token with Dynatrace API and create a new token with 
```

> 'Ingest metrics' scope enabled. 

Reviewer: ```suggestion
'Ingest metrics' (`metrics.ingest`) scope enabled. It is recommended to limit Token scope to only this permission.
```

The endpoint for the Dynatrace Metrics API v2 is 

> * Managed `https://{your-domain}/e/{your-environment-id}/api/v2/metrics/ingest`
> * SaaS `https://{your-environment-id}.live.dynatrace.com/api/v2/metrics/ingest`

Reviewer: ```suggestion
* on Dynatrace Managed: `https://{your-domain}/e/{your-environment-id}/api/v2/metrics/ingest`
* on Dynatrace SaaS: `https://{your-environment-id}.live.dynatrace.com/api/v2/metrics/ingest`
```

```toml
[[outputs.dynatrace]]

>   ## If no OneAgent is running on the host, url and api_token should be set

Reviewer: ```suggestion
  ## If no OneAgent is running on the host, url and api_token need to be set
```

  ## Dynatrace Metrics Ingest v2 endpoint to receive metrics

>   url = "https://YOUR_DOMAIN/api/v2/metrics/ingest"

Reviewer: ```suggestion
  url = "https://{your-environment-id}.live.dynatrace.com/api/v2/metrics/ingest"
```
consistent with the above

  ## API token is required if a URL is specified and should be restricted to the 'Ingest metrics' scope

>   api_token = "your API token here"

Reviewer: ```suggestion
  api_token = "your API token here" // hard-coded for illustration only, should be read from environment
```

```

> You can learn more about how to use the Dynatrace API [here](https://www.dynatrace.com/support/help/dynatrace-api/)

Reviewer: ```suggestion
You can learn more about how to use the Dynatrace API [here](https://www.dynatrace.com/support/help/dynatrace-api/).
```

## Configuration

### `url`

> *required*: `false`

Reviewer: ```suggestion
*required*: `false`  
```
to force a linebreak

Author: I would rather not depend on trailing whitespace

```suggestion
*required*: `false`

```

*default*: Local OneAgent endpoint

> Set Dynatrace environment URL (e.g.: `https://YOUR_DOMAIN/api/v2/metrics/ingest`) if you do not use a OneAgent or wish to export metrics directly to a Dynatrace metrics v2 endpoint. If a URL is set to anything other than the local OneAgent endpoint, then an API token is required.

Reviewer: ```suggestion
Set your Dynatrace environment URL (e.g.: `https://{your-environment-id}.live.dynatrace.com/api/v2/metrics/ingest`) if you do not use a OneAgent or wish to export metrics directly to a Dynatrace metrics v2 endpoint. If a URL is set to anything other than the local OneAgent endpoint, then an API token is required.
```

```toml
url = "https://{your-environment-id}.live.dynatrace.com/api/v2/metrics/ingest"
```

> ### `api_token` - require

Reviewer: ```suggestion
### `api_token`
```

*required*: `false` unless `url` is specified

> API token is required if a URL is specified and should be restricted to the 'Ingest metrics' scope

Reviewer: ```suggestion
API token is required if a URL other than the OneAgent endpoint is specified and it should be restricted to the 'Ingest metrics' scope.
```

```toml
api_token = "your API token here"
```

### `prefix`

*required*: `false`

> 

Reviewer: ```suggestion

Optional prefix to be prepended to all metric names (will be separated with a `.`).

```

```toml

> prefix = "telegraf"

Reviewer 2: In the diff the prefix had a trailing `.` before - do we need to add it?

```suggestion
prefix = "telegraf."
```

Author: No

```

### `insecure_skip_verify`

*required*: `false`

> Setting this option true skips TLS verification

Reviewer: ```suggestion
Setting this option to true skips TLS verification for testing or within local environments.
```

Author: This is also useful in some cases for managed when trying to use self-signed certs in my experience. Do you think we should mention that or leave it alone?

Reviewer: That's what I wanted to imply with local (not necessarily the same machine but also the same data center).

Author: Changed the wording to be more explicit then.

```toml
insecure_skip_verify = false
```

### `additional_counters`

*required*: `false`

If you want to convert values represented as gauges to counters, add the metric names here.

```toml
additional_counters = [ ]
```

### `default_dimensions`

*required*: `false`

> Default dimensions will be added to every exported metric

Reviewer: ```suggestion
Default dimensions that will be added to every exported metric.
```
